overlay/usr/lib/inithooks/bin/nextcloud.py

```diff
@@ -27,59 +38,230 @@ def usage(s=None):
+            - ERR_PASSWORD
+            - ERR_EXCEPTION
+            - ERR_UNKNOWN
+    """
+    local_env = os.environ.copy()
+    local_env["OC_PASS"] = password
+    p = subprocess.run(
+        args=[
+            "/usr/local/bin/turnkey-occ",
+            "user:resetpassword",
+            "--password-from-env",
+            "admin",
+        ],
+        cwd="/var/www/nextcloud",
+        env=local_env,
+        text=True,
+        capture_output=True,
+    )
+    # Success :)
+    if p.returncode == SUCCESS:
+        log.write("admin password set successfully")
+        return (SUCCESS, "", [])
+
+    # password setting failure/error handling
+    else:
+        # occ seems to only output to stdout - even if stacktrace
+        # but just in case that changes
+        occ_output = p.stdout + p.stderr
+        services = []
+        # log raw occ output - including full stacktrace if relevant
+        log.write(occ_output, "err")
+
+        # password complexity failure
+        if "password" in occ_output.lower():
+            error_code = ERR_PASSWORD
+            msg = occ_output.rstrip()
+
+        # nextcloud exception
+        elif "exception" in occ_output.lower():
+            error_code = ERR_EXCEPTION
+            msg = "Nextcloud exception:"
+
+            if "redis" in occ_output.lower():
+                services.append("Redis")
+            if "database" in occ_output.lower():
+                services.append("MariaDB")
+            if services:
+                msg = f"{msg} Can't connect to service(s)"
+            else:
+                msg = f"{msg} Unknown exception"
+
+        # any other unknown/unexpected error
+        else:
+            error_code = ERR_UNKNOWN
+            msg = f"Unexpected Nextcloud error"
+
+        return (error_code, msg, services)
+
+
+def set_random_pass() -> tuple[int, str]:
+    """Returns exit_code & msg"""
+    chars = string.ascii_letters + string.digits
+    while True:
+        random_pass = ''.join(secrets.choice(chars) for _ in range(32))
+        if (
+            any(c.islower() for c in random_pass)
+            and any(c.isupper() for c in random_pass)
+            and sum(c.isdigit() for c in random_pass)
+        ):
+            break
+    while True:
+        exit_code, msg, _ = set_password(random_pass)
+        if exit_code == ERR_PASSWORD:
+            # a 32 char random string should never have this condition, but
+            # if it does, then retrying should resolve it
+            continue
+        else:
+            if exit_code == SUCCESS:
+                with open(RAND_PWD_FILE, "w") as fob:
+                    fob.write(f"{random_pass}\n")
+                msg = f"Admin user random password set: see {RAND_PWD_FILE}"
+                log.write(msg, "warn")
+            else:
+                # all other errors
+                msg = f"Setting password failed - left unset:\n{msg}"
+                log.write(msg, "err")
+            return (exit_code, msg)
```

Comment:
I believe the exit code here is always ignored

Reply:
You were right, but the intention of the exit_code was to handle the corner case of the random password setting fails. It should never happen as other errors are checked for first, but you never know!